### ISSUE TITLE
feature requests

### ISSUE THREAD

**Issue by paulo-erichsen** [arensb/ansible-truenas#7]
Hi @arensb thanks for sharing this ansible collection :)

I'd wondering if these things could be implemented

- [x] Services: [Configuring the SMART Service](https://www.familybrown.org/dokuwiki/doku.php?id=fester112:smart_email)
- [x] Tasks: [Adding SMART Tests](https://www.familybrown.org/dokuwiki/doku.php?id=fester112:smart_schedule)
- [x] Tasks: [Adding Pediodic Snapshot Tasks](https://www.familybrown.org/dokuwiki/doku.php?id=fester112:snapshots)
- [x] Tasks: [Adding Scrub Tasks](https://www.familybrown.org/dokuwiki/doku.php?id=fester112:scrub)

**Comment by arensb** [comment]
The good news is that periodic snapshots are already there:
https://arensb.github.io/truenas/pool_snapshot_task_module.html#ansible-collections-arensb-truenas-pool-snapshot-task-module

I'll see if I can add SMART and scrub stuff, but I don't know when I'll be able to get to this.

In the meantime, maybe you can help me out by thinking about what these modules _should_ look like: how would you like to write an Ansible playbook for them? What should the parameters be called, so as to make sense to the people using them, and also fit into the Ansible ecosystem. Are there any results that it would make sense to return, besides "it succeeded" or "it failed"?

**Comment by paulo-erichsen** [comment]
If I could configure those, I'd probably do something like this

```yaml
---
- name: Configure SMART disk temperature checks
  arensb.truenas.smart:
    check_interval: 30
    power_mode: never
    difference: 0
    informational: 0
    critical_temperature: 45
 
- name: Add SMART short disk periodic tests
  arensb.truenas.pool_smart_task:
    disks: all
    type: short
    description: "run SHORT tests on the 7th, 17th and 27th of each month at 1am"
    schedule: "0 1 7,17,27 * *"
                         
- name: Add SMART long disk periodic tests
  arensb.truenas.pool_smart_task:
    disks: da0,da1,da2,da3
    type: long
    description: "run LONG tests on the 1st and 15th of each month at 1am"
    schedule: "0 1 1,15 * *"
    
- name: Add monthly scrub tasks to the tank pool
  arensb.truenas.pool_scrub_task:
    enabled: true
    pool: tank
    threshold_days: 22
    description: "run Scrub tasks on the 19th of each month at 1am"
    schedule: "0 1 19 * *"
```

**Comment by arensb** [comment]
Thanks. One consideration here is: how can we tell whether a job on the TrueNAS system is the same one as the one defined in an Ansible play? That is, if the play says to run a short SMART test on disk `da0` every Monday, and the TrueNAS system currently has one job: a short SMART test on disk `da0` every Tuesday, does that mean that the existing job should be changed to run on Mondays instead of Tuesdays? Or that we should add a second job, one that runs on Mondays, and ignore the Tuesday job?

I'm pretty sure I had to solve this same problem for NFS mounts or something, but I'm open to other people's ideas.

**Comment by arensb** [comment]
I've added a `smart` module on branch `feature/pr7-smart`. Usage docstrings are at the top of [the module](https://github.com/arensb/ansible-truenas/blob/feature/pr7-smart/plugins/modules/smart.py).

Can you please try it out and see if it works for you?

**Comment by paulo-erichsen** [comment]
> One consideration here is: how can we tell whether a job on the TrueNAS system is the same one as the one defined in an Ansible play?


##  approach 1

I wouldn't mind if the module required users to add the object `id`. The first time users run, if `id` is undefined, then CREATE, but if `id` is defined, then UPDATE.
If this approach was taken, it would be straightforward, but we'd need really good documentation about it, including how to find out the ID such that users can edit their playbooks accordingly.

## approach 2

Another approach without specifying ids would be for the module itself to try to create an "primary key" to find the right object. Example: maybe pool_smart_task could use a primary key of "description" (similar to what was done to the snapshots module). Or maybe the pool_scrub_task could have a primary key of "pool+description"

IMO approach 1 is the more straightforward - it may be more difficult for users since they'd need to specify the ids in their playbooks, but this has the least amount of gotchas

**Comment by paulo-erichsen** [comment]
> I've added a smart module on branch feature/pr7-smart. Usage docstrings are at the top of [the module](https://github.com/arensb/ansible-truenas/blob/feature/pr7-smart/plugins/modules/smart.py).

> Can you please try it out and see if it works for you?

I'll try it out and let you know :)

**Comment by paulo-erichsen** [comment]
Back to my first comment, when choosing between supporting arrays or not supporting arrays, I think it would be simpler if NO arrays were supported.

Looking into some examples how official ansible modules are done, I see this

- https://docs.ansible.com/ansible/latest/collections/ansible/builtin/user_module.html

```
- name: Add the user 'johnd' with a specific uid and a primary group of 'admin'
  ansible.builtin.user:
    name: johnd
    comment: John Doe
    uid: 1040
    group: admin

- name: Add the user 'james' with a bash shell, appending the group 'admins' and 'developers' to the user's groups
  ansible.builtin.user:
    name: james
    shell: /bin/bash
    groups: admins,developers
    append: yes
```

Notice how each user is modified separately instead of through a single call

**Comment by arensb** [comment]
> I wouldn't mind if the module required users to add the object `id`.

From what I've seen, the TrueNAS middleware assigns IDs for you, and doesn't allow you to choose them. So that's not an option.

But it does look similar to something I had to do elsewhere, in the `sharing_nfs` module, I think: there, you have to specify the path to export, and the export flags. So if the flags on the NAS don't match the ones in the Ansible play, what should the module do?

The way I solved it there was to require a description (similar to the "Description" field in S.M.A.R.T. test tasks). For `state == present`, it would look for an export with a matching description.

Also, for scheduled tasks, I'd like to stick to the same parameters as `builtin.cron`, for consistency.

**Comment by paulo-erichsen** [comment]
> From what I've seen, the TrueNAS middleware assigns IDs for you, and doesn't allow you to choose them. So that's not an option.

right. What I was thinking was when the user created the object, the user would need to manually edit the playbook to refer to the ID stored in the database. Example, the user would need to run `midclt call pool.scrub.query | jq`, take note of the IDs and then write the playbook using the IDs. Also when the user created the item through the playbook, the module should return the new ID.

But I understand that this approach can be complicated for users since they would need to know how to lookup the IDs and also modify their playbooks to differentiate between CREATE and UPDATE.

Either way, @arensb what you're doing is great! I could see it going either way. I'd suggest that once you add these and get a new release out, to definitely share them on reddit and discord. (I've found your post on reddit and that's how I ended up here)

**Comment by arensb** [comment]
I've added a new branch, `[feature/pr7-smart-test-task](https://github.com/arensb/ansible-truenas/tree/feature/pr7-smart-test-task)` with a draft `smart_test_task` module. Give it a shot and see if it works.

You'll have to figure out the playbook from the docstring at the top. I haven't finished the examples and output sections.

**Comment by paulo-erichsen** [comment]
> I've added a new branch, [feature/pr7-smart-test-task](https://github.com/arensb/ansible-truenas/tree/feature/pr7-smart-test-task) with a draft smart_test_task module. Give it a shot and see if it works.

Nice! I'll check it out for sure, but it will be a few days until I get to it since I'll be out of town

**Comment by paulo-erichsen** [comment]
@arensb I don't know if you had a chance to see this comment, but this was one issue I had found

# feature/pr7-smart

> Ok, I've tested it a bit and it's able to set all values, except for "interval". It currently seems to ignore the "interval" setting that was passed in.

# feature/pr7-smart-test-task

it seems this also has another issue. My config is this

```
- name: Add SMART short disk periodic tests
  arensb.truenas.smart_test_task:
    name: "run SHORT tests on the 7th, 17th and 27th of each month at 1am"
    disks: ALL
    test: short
    minute: 0
    hour: 1
    day: "7,17,27"
    month: "*"
    weekday: "*"

- name: Add SMART long disk periodic tests
  arensb.truenas.smart_test_task:
    name: "run LONG tests on the 1st and 15th of each month at 1am"
    disks: ada0,ada1
    test: long
    minute: 0
    hour: 1
    day: "1,15"
    month: "*"
    weekday: "*"
```

but when I run it returns this

```
An exception occurred during task execution. To see the full traceback, use -vvv. The error was: KeyError: 'schedule'
fatal: [root@192.168.122.100]: FAILED! => {"changed": false, "module_stderr": "Shared connection to 192.168.122.100 closed.\r\n", "module_stdout": "Traceback (most recent call last):\r\n  File \"/root/.ansible/tmp/ansible-tmp-1693974349.8145535-3600-172060455326188/AnsiballZ_smart_test_task.py\", line 107, in <module>\r\n    _ansiballz_main()\r\n  File \"/root/.ansible/tmp/ansible-tmp-1693974349.8145535-3600-172060455326188/AnsiballZ_smart_test_task.py\", line 99, in _ansiballz_main\r\n    invoke_module(zipped_mod, temp_path, ANSIBALLZ_PARAMS)\r\n  File \"/root/.ansible/tmp/ansible-tmp-1693974349.8145535-3600-172060455326188/AnsiballZ_smart_test_task.py\", line 47, in invoke_module\r\n    runpy.run_module(mod_name='ansible_collections.arensb.truenas.plugins.modules.smart_test_task', init_globals=dict(_module_fqn='ansible_collections.arensb.truenas.plugins.modules.smart_test_task', _modlib_path=modlib_path),\r\n  File \"/usr/local/lib/python3.9/runpy.py\", line 225, in run_module\r\n    return _run_module_code(code, init_globals, run_name, mod_spec)\r\n  File \"/usr/local/lib/python3.9/runpy.py\", line 97, in _run_module_code\r\n    _run_code(code, mod_globals, init_globals,\r\n  File \"/usr/local/lib/python3.9/runpy.py\", line 87, in _run_code\r\n    exec(code, run_globals)\r\n  File \"/tmp/ansible_arensb.truenas.smart_test_task_payload_r380z7wa/ansible_arensb.truenas.smart_test_task_payload.zip/ansible_collections/arensb/truenas/plugins/modules/smart_test_task.py\", line 382, in <module>\r\n  File \"/tmp/ansible_arensb.truenas.smart_test_task_payload_r380z7wa/ansible_arensb.truenas.smart_test_task_payload.zip/ansible_collections/arensb/truenas/plugins/modules/smart_test_task.py\", line 237, in main\r\nKeyError: 'schedule'\r\n", "msg": "MODULE FAILURE\nSee stdout/stderr for the exact error", "rc": 1}
```

this was fixed with adding this line

```diff
# Collect arguments to pass to resource.create()
arg = {
    'desc': name,
+    'schedule': {},
}
```

**Comment by arensb** [comment]
I was able to reproduce the bug in feature/pr7-smart-test-task, and fix it. As you suspected, it had to do with a messed-up `schedule` field. (Or, more precisely, I found a bug in the schedule-processing code while editing the "if the resource exists" section, but then didn't apply the fix correctly to the "if we need to create the resource" section.) Please give it another try.

> IMO the ID that ansible uses in smart_test_task.py may not be best to use description since the UI doesn't even require that parameter. I think it would be better if it was possible to make the ID based on DISKS+TYPE, but that has it's own cons

It looks as though the required fields to create a S.M.A.R.T. Test task are the test type ("LONG", "SHORT", etc.) and either `all_disks` or a list of disks. However, that does not uniquely identify a job: you can have several `LONG` jobs on disk `ada0`.

In fact, there may be valid reasons for doing this: maybe you have several TrueNAS machines, and you want to run a `SHORT` test on `ada0,ada1` on all machines on the 1st of each month. But some of them are backup servers that keep user data on those disks, and you want to test them on the 15th of each month as well:

    - hosts: truenas-hosts
      tasks:
        - arensb.truenas.smart_test_task:
            disks: [ ada0, ada1 ]
            test: short
            day: "1"
    - hosts: backup-servers
      tasks:
        - arensb.truenas.smart_test_task:
            disks: [ ada0, ada1 ]
            test: short
            day: "15"

Assuming that `backup-servers` is a subset of `truenas-hosts`, we now have some hosts with two jobs with the same disks and type.

The purpose of an identifier is to be able to point to something (a task, in this case) and say "there: that's the one I'm talking about". That seems easier to do with the description field than with any other field or combination of fields. Yes, it means that you can't change the description without deleting and recreating the task, but hopefully that won't happen too often.

A bigger problem, to my mind, is that this module doesn't allow you to delete _all_ tasks and reset the machine to a known state where the only tasks are the one defined in your playbook.

**Comment by paulo-erichsen** [comment]
> That seems easier to do with the description field than with any other field or combination of fields. Yes, it means that you can't change the description without deleting and recreating the task, but hopefully that won't happen too often.

Agreed. that seems the most simple and straightforward way to handle this

> [I fixed it,] please give it another try.

will do :)

**Comment by arensb** [comment]
> DONE - looks good now

Thanks. I think that just leaves the [S.M.A.R.T. service](https://github.com/arensb/ansible-truenas/blob/feature/pr7-smart/plugins/modules/smart.py) itself. Have you had a chance to try that out?

**Comment by paulo-erichsen** [comment]
I did. The updated change that fixes the interval works! :heavy_check_mark:

**Comment by paulo-erichsen** [comment]
while I have your attention @arensb, I've opened another much smaller request #8

**Comment by paulo-erichsen** [comment]
closed with v1.8.0
